File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var (
	min  = math.MaxFloat64
	max  = math.MaxFloat64 * (-1)
	num  int
	nums []float64
)

func dfs(depth int, res float64, plus, minus, multiply, divide int) {
	if depth == num {
		min = math.Min(res, min)
		max = math.Max(res, max)
		return
	}
	if plus > 0 {
		dfs(depth+1, res+nums[depth], plus-1, minus, multiply, divide)
	}
	if minus > 0 {
		dfs(depth+1, res-nums[depth], plus, minus-1, multiply, divide)
	}
	if multiply > 0 {
		dfs(depth+1, res*nums[depth], plus, minus, multiply-1, divide)
	}
	if divide > 0 {
		dfs(depth+1, float64(int(res/nums[depth])), plus, minus, multiply, divide-1)
	}
}
func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	fmt.Fscanf(reader, "%d\n", &num)
	nums = make([]float64, num)
	var methods = make([]int, 4)
	for i := 0; i < num; i++ {
		fmt.Fscanf(reader, "%f ", &nums[i])
	}
	for i := 0; i < 4; i++ {
		fmt.Fscanf(reader, "%d ", &methods[i])
	}
	dfs(1, nums[0], methods[0], methods[1], methods[2], methods[3])
	fmt.Fprintf(writer, "%0.f\n%0.f", max, min)
}

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
min_num = float('inf')
max_num = -float('inf')
num = int(input())
nums= list(map(int, input().split()))
methods = list(map(int, input().split()))

def dfs(depth, res, plus, minus, multiple, divide):
    global min_num, max_num
    if depth==num:
        min_num=min(min_num, res)
        max_num=max(max_num, res)
        return
    
    if plus:
        dfs(depth+1, res+nums[depth], plus-1, minus, multiple, divide)
    if minus:
        dfs(depth+1, res-nums[depth], plus, minus-1, multiple, divide)
    if multiple:
        dfs(depth+1, res*nums[depth], plus, minus, multiple-1, divide)
    if divide:
        dfs(depth+1, int(res/nums[depth]), plus, minus, multiple, divide-1)

dfs(1, nums[0], methods[0], methods[1], methods[2], methods[3])

print(max_num)
print(min_num)